Add CONTRIBUTING.md (#3415)
rancher/rke: CONTRIBUTING.md

# Contributing to RKE #

Thanks for taking the time to contribute to RKE!

> Of course, contributing code is more than welcome! To keep things simple, if you're fixing a small issue, you can simply submit a PR and we will pick it up. However, if you're planning to submit a bigger PR to implement a new feature or fix a relatively complex bug, please continue reading.

Reviewer: > To keep things simple, if you're fixing a small issue, you can simply submit a PR and we will pick it up.
2 things:
1. I would still prefer to have an issue to track even "small issues". The only exceptions would be things like typos / misc changes. Anything that fixes an actual issue or implements a feature should an issue IMO.
2. I would remove "we will pick it up" as it establishes commitment to review any submitted PR.

Author: 1. Yes, let's change that. The idea was to facilitate users contributing. Having to create an issue is another step which could be perceived as a burden. But as we don't expect RKE to become a "popular" open source project like k3s, we can remove that
2. I believe we should commit to review any submitted PR. Why not?

### Opening PRs and organizing commits
PRs should generally address only 1 issue at a time. If you need to fix two bugs, open two separate PRs. This will keep the scope of your pull requests smaller and allow them to be reviewed and merged more quickly.

> Each PR needs to be linked to a Github issue in [rancher project](https://github.com/rancher/rancher) which provides as much detail as possible on the commit at its motivation. If it is a bug fix, please describe how to reproduce it in the issue. Make the team or Project Manager aware of the Github issue so that they can set the correct milestone. Milestones are very important for release and QA tracking.

Reviewer 2: 1. Can we create a PR template for rancher/rke? It'd be a lot easier to direct contributors on what/what is required to add to a PR.
2. As I understand it, we are focusing much more on `what code has been modified/added` in the PR description to be able to catch regressions quicker, not just adding repro steps to the issue. I'd add a splurge about that to make that obvious to contributors. Our focus on regression prevention needs to extend to community PRs too.

Author: re: 1, yes, we could do that but in another PR. Could you create an issue about it?

re: 2. if you are referring to the line `If it is a bug fix, please describe how to reproduce it in the issue`. The reason is mainly that we should triage the issue to verify that it is indeed a general RKE1 issue. A lot of users have problems which end up being a misconfiguration in their hosts

Reviewer 2: 1. Yes! https://github.com/rancher/rke/issues/3442
2. True, but that's an addition to the issue and this section is on 'Opening PRs and organizing commits' so I think it'd be helpful to add docs that will help the contributor focus on _what code/additions were added_ in the PR itself.

Author: I change opinion :). I think it is best to add the PR Template to this PR because it is related

> Each PR needs to be linked to a Github issue in [rancher project](https://github.com/rancher/rancher) which provides as much detail as possible on the commit at its motivation. If it is a bug fix, please describe how to reproduce it in the issue. Make the team or Project Manager aware of the Github issue so that they can set the correct milestone. Milestones are very important for release and QA tracking.

Reviewer: Minor wording changes, but I agree with @a-blender 's suggestion to establish PR template.

```suggestion
Each PR needs to be linked to a Github issue in [rancher project](https://github.com/rancher/rancher) which provides as much detail as possible on the change and its motivation. If it is a bug fix, please describe how to reproduce it in the issue. Make the Team or Project Manager aware of the GitHub issue so that they can set the correct milestone. Milestones are very important for the release and QA tracking.
```

Author: PR template proposal now included in this PR

Generally, pull requests should consist of a single logical commit. However, if your PR is for a large feature, you may need a more logical breakdown of commits. This is fine as long as each commit is a single logical unit. Changes generated by `go mod tidy` or `go generate` should go in separate commits.

Pull requests should target the release/vX.Y branch instead of master. The latest code in RKE is included in such release branches.

### Writing good commit messages

> Git commit messages should explain the how and why of your change and be separated into a brief subject line followed by a more detailed body. When in doubt, follow this guide for good commit messages and you can’t go wrong: https://chris.beams.io/posts/git-commit/.

Reviewer: I don't think we generally following this approach and using PR description in lieu of "detailed body". Our commit messages are generally short - for better or worse.

Author: ok, I can change that. I don't care where things are described as long as they are described :)

One particularly useful point made in the above guide is regarding commit subject lines:

> A properly formed Git commit subject line should always be able to complete the following sentence:
> 
> - If applied, this commit will <ins>your subject line here</ins>

A simple but effective convention to follow for commits is the “problem / solution” pattern. It looks like this:
```
<Subject>

Problem: <Statement of problem>

Solution: <Statement of solution>
```

As an example, here is a commit taken from the rancher/rancher repo:
```
commit b71ce2892eecb7c87a5212e3486f1de899a694aa
Author: Dan Ramich <danold215@gmail.com>
Date:   Tue Jun 19 11:56:52 2018 -0700

    Add Validator for RoleTemplate

    Problem:
    Builtin RoleTemplates can be updated through the API

    Solution:
    Add a Validator to ensure the only field that can be changed on a
    builtin RoleTemplate is 'locked'
```

### Reviewing, addressing feedback, and merging
Generally, pull requests need two approvals from maintainers to be merged.

> When addressing review feedback, it is helpful to the reviewer if additional changes are made in new commits. This allows the reviewer to easily see the delta between what they previously reviewed and the changes you added to address their feedback. Once positively reviewd, those commits should be squashed into one.

Reviewer: ```suggestion
When addressing review feedback, it is helpful to the reviewer if additional changes are made in new commits. This allows the reviewer to easily see the delta between what they previously reviewed and the changes you added to address their feedback. Once positively reviewed, those commits should be squashed into one.
```

Once a PR has at least two approvals, it can be merged. Here’s how the merge should be handled:

> - If the PR is a single logical commit, the merger should use the “Rebase and merge” option. This keeps the git commit history very clean and simple and eliminates noise from "merge commits."

Reviewer 2: Are the two first points backwards? If a PR has a single commit with added/updated logic `Create a merge commit` and if 1+ logic commits then `Rebase and merge`. Will also wait for @superseb's comments to be addressed and re-review this section after that.

Author: Good catch! I am missing the part where the commits should be squashed together once the PR has been approved

> - If the PR is more than one logical commit, the merger should use the “Create a merge commit” option.

Reviewer: Suggesting to specifically call out the case of `go generate` commits here so they are not accidentally squashed.

Author: good point

- If the PR consists of more than one commit because the author added commits to address feedback, the commits should be squashed into a single commit (or more than one logical commit, if it is a big feature that needs more commits). This can be achieved in one of two ways:
  - The merger can use the “Squash and merge” option. If they do this, the merger is responsible for cleaning up the commit message according to the previously stated commit message guidance.
  - The pull request author, after getting the requisite approvals, can reorganize the commits as they see fit (using, for example, git rebase -i) and re-push.

Once the PR is merged, please change the status of the linked issue in [rancher project](https://github.com/rancher/rancher) to "Needs QA". In the following days, the QA team will run tests and process the PR to verify that no regressions were introduced. Please be alert on the issue as the QA team might ask questions related to the PR. 

### Backports

#### RKE code

> To understand how backports should be pulled through, RKE project releases and what is included in each release should be described. RKE in general maintains two tracks in parallel which are differentiated by the minor version. Each of these tracks are developped in two different github branches:

Reviewer 2: `release/v1.3` is not being actively maintained but will have critical bugs backported to it and released again if it's necessary to do so. It's not EOL.

Author: Thanks for this information, I was not aware of it. When does a version become EOL?

Reviewer 2: Rancher uses rke under the hood so it's tied to Rancher and the EOL for supported Kubernetes versions. For example, the `release/v1.3` line corresponds with Rancher 2.6 clusters and according to the support matrix, https://www.suse.com/suse-rancher/support-matrix/all-supported-versions/rancher-v2-6-13/ the latest Rancher 2.6 only supports up to k8s 1.24 clusters.

We are far beyond that now with k8s 1.26 reaching active EOL (and the maintenance phase) on Feb 28, 2024 https://kubernetes.io/releases/. Customers using rke 1.3.x are most likely in a position to upgrade than ask for new features or enhancements on their current version of rke. That being said, we still need to support customers on old versions so they still require passive maintenance.

cc @kinarashah do we have any rke versions that are post-maintenance EOL?

Author: My understanding was that once k8s upstream version reaches EOL, we don't do any support to our customers or maybe just best effort. This will change when we announce the LTS kubernetes product which I don't know when will it happen

Author: My understanding was that once k8s upstream version reaches EOL, we don't do any support to our customers or maybe just best effort. This will change when we announce the LTS kubernetes product which I don't know when will it happen

> To understand how backports should be pulled through, RKE project releases and what is included in each release should be described. RKE in general maintains two tracks in parallel which are differentiated by the minor version. Each of these tracks are developped in two different github branches:

Reviewer: Minor typos
```suggestion
To understand how backports should be pulled through, RKE project releases and what is included in each release should be described. RKE in general maintains two tracks in parallel which are differentiated by the minor version. Each of these tracks is developed in two different GitHub branches:
```

* release/v1.x, (for example [release/v1.5](https://github.com/rancher/rke/tree/release/v1.5)
* release/v1.(x-1) (for example [release/v1.4](https://github.com/rancher/rke/tree/release/v1.4)

> Each track releases at the same time an RKE binary. Each release of RKE officially supports a few Kubernetes versions. Some versions are shared by both tracks but normally the track with the higher minor version includes at least one newer Kubernetes version. For example, as it can be observed in [release v1.4.11](https://github.com/rancher/rke/releases/tag/v1.4.11), the following Kubernetes versions are supported:

Reviewer 2: Awk wording. Suggest: _Each track releases an RKE binary._

* v1.26.9-rancher1-1
* v1.25.14-rancher1-1
* v1.24.17-rancher1-1
* v1.23.16-rancher2-3

whereas [release v1.5.0](https://github.com/rancher/rke/releases/tag/v1.5.0) includes:
* v1.25.14-rancher1-1
* v1.26.9-rancher1-1
* v1.27.6-rancher1-1

> By support, it means we can deploy a cluster with that Kubernetes version using the binray included in that RKE release.

Reviewer 2: Suggest: _If a Kubernetes version is **supported**, that means we can deploy a cluster with that Kubernetes version using the binary included in that RKE release._

* Create an Issue in rancher/rancher github project for Project Managers and QA to track the commit. Set the decided milestone for release/v1.x in the issue
* Create another Issue in the same github project and in the description, specify it is a backport of the previous commit. Set the decided milestone for release/v1.(x-1) in the issue
* Create the PR with the commit in branch release/v1.x and in the description, link the newly created issue
* Cherry-pick the commit in branch release/v1.(x-1) and in the description, link the created issue for that RKE track.

A commit is normally available in all supported Kubernetes versions of the release unless scoped in the code to a specific version. Here are two examples for this:
* Available in all supported Kubernetes versions: https://github.com/rancher/rke/commit/1dad2582e1efe5564b6d1a7385afad70d5dbd1f7
* Scoped to a specific version: https://github.com/rancher/rke/commit/feca6a3048c1f8c9031ab2849a46c46a7ea1d673

> Note that in the scoped case, there is a check on the version, which is normally fetch by executing the line `util.StrToSemVer(k8sVersion)`. In this case, as the same code is executed for all Kubernetes versions, unless scoped, there is no need to create extra PRs to effectively backport the code to all Kubernetes versions.

Reviewer 2: I'd add a link to the code that checks the version

Author: You mean github.com/coreos/go-semver/semver? We are using that type and call functions such as https://github.com/rancher/rke/blob/release/v1.5/cluster/validation.go#L752

Reviewer 2: Or this https://github.com/rancher/rke/blob/c56fb8c42a079d569dd402c1243a5ae16b4004c9/cluster/validation.go#L748 but yes

#### Add-ons

RKE consumes different add-ons to deploy a usable Kubernetes cluster, e.g. CNI plugins. Typically all Kubernetes versions include the same range of add-ons but use different add-on versions. For example, for the flannel add-on:
* In v1.26.8-rancher1-1 flannel version is `v0.21.4`
* In v1.24.17-rancher1-1 flannel version is `v0.15.1`

> What add-on version is used per RKE Kubernetes release is mapped in the `data.json` file, which is a file autogenerated by querying [KDM](https://github.com/rancher/kontainer-driver-metadata) releases. Read the RKE [README.md](https://github.com/rancher/rke/blob/release/v1.4/README.md) for more information.

Reviewer: Suggesting to use relative link so we're not pointing to specific branch (default branch is already `release/v1.5` while this URL points to `release/v1.4`).

Author: What do you mean by relative link? I thought I always need to specify the release in the url to access the file (README.md in this case)

Reviewer 3: Could you link this to templates.go file that maps the versions to k8s versions instead?

# Contributing to RKE #

Thanks for taking the time to contribute to RKE!

Of course, contributing code is more than welcome! To keep things simple, if you're fixing a small issue, you can simply submit a PR and we will pick it up. However, if you're planning to submit a bigger PR to implement a new feature or fix a relatively complex bug, please open an issue that explains the change and the motivation for it. If you're addressing a bug, please explain how to reproduce it.

### Opening PRs and organizing commits
PRs should generally address only 1 issue at a time. If you need to fix two bugs, open two separate PRs. This will keep the scope of your pull requests smaller and allow them to be reviewed and merged more quickly.

When possible, fill out as much detail in the pull request template as is reasonable. Most important is to reference the GitHub issue that you are addressing with the PR.

Generally, pull requests should consist of a single logical commit. However, if your PR is for a large feature, you may need a more logical breakdown of commits. This is fine as long as each commit is a single logical unit.

> 

Reviewer 4: I think we should mention that generally `go generate` and ` go mod tidy` outputs should be in dedicated commits

Author: Good point Thomas,thanks! Let me add that

Once a PR has the necessary approvals, it can be merged. Here’s how the merge should be handled:
- If the PR is a single logical commit, the merger should use the “Rebase and merge” option. This keeps the git commit history very clean and simple and eliminates noise from "merge commits."
- If the PR is more than one logical commit, the merger should use the “Create a merge commit” option.
- If the PR consists of more than one commit because the author added commits to address feedback, the commits should be squashed into a single commit (or more than one logical commit, if it is a big feature that needs more commits). This can be achieved in one of two ways:
  - The merger can use the “Squash and merge” option. If they do this, the merger is responsible for cleaning up the commit message according to the previously stated commit message guidance.
  - The pull request author, after getting the requisite approvals, can reorganize the commits as they see fit (using, for example, git rebase -i) and re-push.

> ### Backports

Reviewer 5: This is not enough context for a contributor. `A commit` can be core RKE code that applies to everything, which could be made k8s version (minor/patch) specific or only applies to a subset like an add-on which is already k8s version specific. This needs to be split up in the bigger currently known changes we have and address those.

Author: Good point but I don't get what you mean by `needs to be split up in the bigger currently known changes we have and address those`. I'll try to rewrite it to cover the different backport cases

Reviewer 5: We know what kind of changes we have, these need to be described in a bit more detail so the reader can identify what change falls under what category. If the reader can do that, the reader can also identify where it needs to be backported.

We have RKE code backports, RKE k8s version backports and add-on templates backports at minimum which all have their own "policy" or at least it needs to be described.

Author: This is what I have written.
RKE code backports:
```
By design, RKE code supports all active Kubernetes minor versions, so the commit in the RKE project is automatically available for all. You just need to remember to contribute the commit in all the active RKE releases which are developed in `release vX.Y` branches
```
add-on templates backports:
```
if the commit is done in a dependent project, such as an add-on, the `data.json` file will need to be updated accordingly, so that all Kubernetes minor versions supported by RKE consume the add-on version that includes the commit
```
Could you elaborate more about the `RKE k8s version backports` case? I am probably missing that one but I'm not sure where is the commit in that case. Thanks Seb!

Reviewer 5: I think this section needs to be divided in categories of changes, that we can identify and refer to in a way that is clear enough that we all know what is what. Sentences like `A commit should be available in all active Kubernetes minor versions` do not clarify enough on what the goal is.

This needs a bit of thinking to get everything cleared up, I've started a bit below:

- RKE code change is, unless scoped in the code to specific Kubernetes versions, available generically. For example, better/more/less logging statements can be merged and will be available to every invocation of RKE. Kubernetes versions are not related in this way because new RKE code gets executed wether you have a current-2, current-1 or current patch and or minor version). Backporting this means copying the code from `release/v1.X` to `release/v1.X-1`, and will be available to every invocation of the new RKE code for `release/v1.X-1`.
- RKE code changes that are scoped to Kubernetes versions/Kubernetes component versions. See https://github.com/rancher/rke/blob/645e80840596a8616c2a1f35cea0c655ac700fa9/cluster/plan.go#L60 as an example. This is outside of KDM.

Both changes above should be backported if they are related to versions that are supported in both `release/v1.X` and `release/v1.X-1` because we don't want users to have differences between provisioning a cluster in `release/v1.X-1` or `release/v1.X`. (If a flag is changed between versions that is only available in one version, an edit to the cluster will trigger provisioning (makes sense) but it will then also replace the component with the newly added flag to be removed if its not in `release/v1.X` (or vice versa)

This is just a beginning, and we don't need to flesh it out completely but this is the basics that we need to nail down to have a starting point. I am also fine with starting with what we have currently but I don't think its a good enough starting point, I'd rather take it a step further and start with more than this.

I also prefer having more clear statements than `By active we mean minor versions that Kubernetes upstream is actively working on, which normally means 3 kubernetes minors.`. We should make a statement now and say that we do latest supported - 2 for example. That is already way more clear than the previous statement. We can fine tune from there but we need some more clarity on these topics.

Lastly, the way things are described need to be more clear. Sentence like `However, if the commit is done in a dependent project, such as an add-on, the `data.json` file will need to be updated accordingly,` is weird because we don't update `data.json`. We updated source files in KDM and generate the `data.json`. I know this sounds nit picky, but for me its the difference between having a lot of confusion between multiple readers and having clarity (saving time because its more clear)

Hope it makes sense and let me know if you get what I mean or not, we can also have a session to go over it.

Author: Thanks for the explanation. My intention was not to cover all the different options in detail as then the document will probably become too long. My experience tells me that too long documents have the danger that nobody ever reads them, thus making this PR a bit useless. But let me try to cover most or all backport use cases and see what we get

Author: I made an update going more into detail and with example. The result is not as long as I anticipated :)

# Contributing to RKE #

Thanks for taking the time to contribute to RKE!

### Opening PRs and organizing commits
PRs should generally address only 1 issue at a time. If you need to fix two bugs, open two separate PRs. This will keep the scope of your pull requests smaller and allow them to be reviewed and merged more quickly.

> Each PR needs to be linked to a Github issue in [rancher project](https://github.com/rancher/rancher) which provides as much detail as possible on the change and its motivation. If it is a bug fix, please describe how to reproduce it in the issue. Make the Team or Project Manager aware of the GitHub issue so that they can set the correct milestone. Milestones are very important for the release and QA tracking. Milestones follow the convention: `$RANCHERVERSION-KDM-$MONTH-$YEAR-patches, e.g. `v2.7-KDM-Nov-2023-patches`.

Reviewer 6: AFAIK, there are multiple formats of the milestones. I suggest removing the last sentence.

Once a PR has at least two approvals, it can be merged. Here’s how the merge should be handled:
- If the PR is a single logical commit, the merger should use the “Rebase and merge” option. This keeps the git commit history very clean and simple and eliminates noise from "merge commits."
- If the PR is more than one logical commit, the merger should use the “Create a merge commit” option.
- If the PR consists of more than one commit because the author added commits to address feedback, the commits should be squashed into a single commit (or more than one logical commit, if it is a big feature that needs more commits). Remember that `go mod tidy` or `go generate` should go in separate commits, hence do not swaush in that case. Squashing of commits can be achieved in one of two ways:
  - The merger can use the “Squash and merge” option. If they do this, the merger is responsible for cleaning up the commit message according to the previously stated commit message guidance.
  - The pull request author, after getting the requisite approvals, can reorganize the commits as they see fit (using, for example, git rebase -i) and re-push.

> Once the PR is merged, please change the status of the linked issue in [rancher project](https://github.com/rancher/rancher) to "Needs QA". In the following days, the QA team will run tests and process the PR to verify that no regressions were introduced. Please be alert on the issue as the QA team might ask questions related to the PR. 

Reviewer 3: Thought we also need to move the issue to `To-Test`, is that no longer applicable?

Author: Ah good point! It would be [zube] To Test, right?

> Once the PR is merged, please change the status of the linked issue in [rancher project](https://github.com/rancher/rancher) to "[zube] To Test". In the following days, the QA team will run tests and process the PR to verify that no regressions were introduced. Please be alert on the issue as the QA team might ask questions related to the PR. 

Reviewer 6: The label "[zube] To Test" is deprecated. Now we change the status of the issue to "Ready for QA" (or similar state) in the Project(s) that the issue is associated with.  GitHub UI shows the Projects section under the Labels section on the right of the page.

### Backports

#### RKE code

> To understand how backports should be pulled through, RKE project releases and what is included in each release should be described. RKE in general maintains two tracks in parallel which are differentiated by the minor version. Each of these tracks s developped in two different GitHub branches:

Reviewer 3: nit typo: `these tracks s developped`

> To understand how backports should be pulled through, RKE project releases and what is included in each release should be described. RKE in general maintains two tracks in parallel which are differentiated by the minor version. Each of these tracks are developped in two different GitHub branches:

Reviewer 6: Could you rephrase the first sentence which is confusing to read (at leat to me)?

* release/v1.x, (for example [release/v1.5](https://github.com/rancher/rke/tree/release/v1.5)
* release/v1.(x-1) (for example [release/v1.4](https://github.com/rancher/rke/tree/release/v1.4)

Each track releases an RKE binary. Each release of RKE officially supports a few Kubernetes versions. Some versions are shared by both tracks but normally the track with the higher minor version includes at least one newer Kubernetes version. For example, as it can be observed in [release v1.4.11](https://github.com/rancher/rke/releases/tag/v1.4.11), the following Kubernetes versions are supported:
* v1.26.9-rancher1-1
* v1.25.14-rancher1-1
* v1.24.17-rancher1-1
* v1.23.16-rancher2-3

whereas [release v1.5.0](https://github.com/rancher/rke/releases/tag/v1.5.0) includes:
* v1.25.14-rancher1-1
* v1.26.9-rancher1-1
* v1.27.6-rancher1-1

If a Kubernetes version is supported, that means we can deploy a cluster with that Kubernetes version using the binary included in that RKE release.

> When doing a commit in RKE code, we must carefully think if we want that commit in both tracks and hence backported. Sometimes, there is a reason to no backport it, for example a flag, that we would like to limit to only the newer track. However, that is exceptional and normally commits end up in both branches as a way to reduce the differences between those and simplify RKE maintainability. Therefore, usually we backport a commit between both RKE tracks. In that case, the following steps should be applied:

Reviewer 3: Sometimes there isn't a reason to backport it/ Sometimes there is a reason not to backport it

> When doing a commit in RKE code, we must carefully think if we want that commit in both tracks and hence backported. Sometimes, there is a reason to no backport it, for example a flag, that we would like to limit to only the newer track. However, that is exceptional and normally commits end up in both branches as a way to reduce the differences between those and simplify RKE maintainability. Therefore, usually we backport a commit between both RKE tracks. In that case, the following steps should be applied:

Reviewer 6: ```suggestion
When doing a commit in RKE code, we must carefully think if we want that commit in both tracks and hence backported. Sometimes, there is a reason not to backport it, for example, a flag that we would like to limit to only the newer track. However, that is exceptional and normally commits end up in both branches as a way to reduce the differences between those and simplify RKE maintainability. Therefore, usually we backport a commit between both RKE tracks. In that case, the following steps should be applied:
```

> * Create an Issue in rancher/rancher github project for Project Managers and QA to track the commit. Set the decided milestone for release/v1.x in the issue

Reviewer 3: Are we now creating `release/v1.x` milestone under rancher/rancher? We usually create this milestone under rke.

Author: If I understood correctly:
* milestones are created in rancher/rancher
* branch `releases/v1.x` is created in rancher/rke project
And they are connected. For example:
* milestone [v2.8-KDM-Nov-2023-patches](https://github.com/rancher/rancher/issues?q=is%3Aopen+is%3Aissue+milestone%3Av2.8-KDM-Nov-2023-patches) in rancher/rancher maps to branch [release/v1.5](https://github.com/rancher/rke/tree/release/v1.5) in rancher/rke
* milestone [v2.7-KDM-Nov-2023-patches](https://github.com/rancher/rancher/issues?q=is%3Aopen+is%3Aissue+milestone%3Av2.7-KDM-Nov-2023-patches) in rancher/rancher maps to branch [release/v1.4](https://github.com/rancher/rke/tree/release/v1.4) in rancher/rke
right?

Author: Assuming what I wrote is correct, let me add one extra sentence to explain this. I can see how it could be confusing for the reader

* Create another Issue in the same github project and in the description, specify it is a backport of the previous commit. Set the decided milestone for release/v1.(x-1) in the issue
* Create the PR with the commit in branch release/v1.x and in the description, link the newly created issue
* Cherry-pick the commit in branch release/v1.(x-1) and in the description, link the created issue for that RKE track.

A commit is normally available in all supported Kubernetes versions of the release unless scoped in the code to a specific version. Here are two examples for this:
* Available in all supported Kubernetes versions: https://github.com/rancher/rke/commit/1dad2582e1efe5564b6d1a7385afad70d5dbd1f7
* Scoped to a specific version: https://github.com/rancher/rke/commit/feca6a3048c1f8c9031ab2849a46c46a7ea1d673

> Note that in the scoped case, there is a check on the version, which is normally fetch by executing the line `util.StrToSemVer(k8sVersion)`, [here an example](https://github.com/rancher/rke/blob/release/v1.5/cluster/validation.go#L752). In this case, as the same code is executed for all Kubernetes versions, unless scoped, there is no need to create extra PRs to effectively backport the code to all Kubernetes versions.

Reviewer 3: > In this case, as the same code is executed for all Kubernetes versions, unless scoped, there is no need to create extra PRs to effectively backport the code to all Kubernetes versions. 

This line needs to be modified, backporting for rke1 is decided based on rke1 versions and once backported - it'll apply to k8s versions it supports. Branches aren't based on k8s versions.

Author: I was trying to explain that there is no need to backport stuff to different k8s versions (as we do in rke2/k3s) because the PR will apply to k8s versions supported by the rke1 version. Let me rephrase it a bit better
